drivers/mcp23x17/Kconfig

```diff
@@ -0,0 +1,40 @@
+# Copyright (c) 2021 Gunar Schorcht
```

Comment:
@leandrolanzieri How can I test the config for a new driver. Everytime I use `menuconfig` as make target, I only see `LM75A/TMP1075` in `Sensor Device Drivers`. When I use `[A] Toggle show-all mode`, I see all variables in red, but I can't change anything. Furthermore, I get the following message in red.
```
*** !! ERROR: There are conflicting modules active !! ***
```

Reply:
@gschorcht to test the dependency modelling with Kconfig you should set a variable: `env TEST_KCONFIG=1 make menuconfig`. 

>Furthermore, I get the following message in red.
>`*** !! ERROR: There are conflicting modules active !! ***`

Everything that is in red is not visible, so that message is actually hidden if I understood you correctly.

For this modelling I'd suggest the following change:

```diff
diff --git a/drivers/mcp23x17/Kconfig b/drivers/mcp23x17/Kconfig
index ca79fcc382..705780263b 100644
--- a/drivers/mcp23x17/Kconfig
+++ b/drivers/mcp23x17/Kconfig
@@ -36,6 +36,8 @@ config MODULE_MCP23X17_IRQ
     depends on HAS_PERIPH_GPIO_IRQ
     depends on MODULE_MCP23X17
     select MODULE_PERIPH_GPIO_IRQ
-    select MODULE_EVENT_THREAD_MEDIUM
+    depends on MODULE_EVENT_THREAD_MEDIUM
+    help
+        To use the IRQs the MODULE_EVENT_THREAD symbol should be set.
 
 endif # MODULE_MCP23X17
diff --git a/tests/driver_mcp23x17/app.config.test b/tests/driver_mcp23x17/app.config.test
index 181cba5f04..9d5911c6bf 100644
--- a/tests/driver_mcp23x17/app.config.test
+++ b/tests/driver_mcp23x17/app.config.test
@@ -2,6 +2,10 @@
 # application configuration. This is only needed during migration.
 CONFIG_MODULE_MCP23X17=y
 
+# enable event thread to process IRQs
+CONFIG_MODULE_EVENT=y
+CONFIG_MODULE_EVENT_THREAD=y
+CONFIG_MODULE_EVENT_THREAD_MEDIUM=y
 # enable interrupt support by default
 CONFIG_MODULE_MCP23X17_IRQ=y
```

You may also want to expose which thread priority to use on the driver, similar to the `at` driver (see [here](https://github.com/RIOT-OS/RIOT/blob/9256970517798bd844f4ee1dc4cddd9f61cef6e1/drivers/at/Kconfig#L22-L48)).

Reply:
@leandrolanzieri Either I still don't understand how Kconfig is used correctly or there is something wrong with ESP8266 Kconfig. After making the changes according to your suggestions and running Kconfig, I get the following output for `USEMODULE` variable:
```
TEST_KCONFIG=1 make -C tests/driver_mcp23x17/ BOARD=esp8266-esp-12x menuconfig
TEST_KCONFIG=1 make -C tests/driver_mcp23x17/ BOARD=esp8266-esp-12x info-debug-variable-USEMODULE
```suggestion
benchmark tsrb core_init periph_init cpu periph_init_uart newlib stdin mcp23x17_irq event core_panic mcp23x17 periph_i2c periph_init_gpio periph_timer mcp23x17_i2c core periph_uart periph_init_pm event_thread periph_common event_thread_medium stdio_uart_rx shell periph_init_timer core_msg periph_gpio_irq newlib_syscalls_default auto_init core_thread_flags periph_gpio periph_pm div sys periph_init_i2c auto_init_xtimer stdio_uart isrpipe board xtimer core_idle_thread periph_init_gpio_irq mcp23x17_irq_medium
```
When I set `USEMODULE` as environment variable, I get the following output for `USEMODULE` variable:
```
USEMODULE='mcp23x17_i2c mcp23x17_irq' make -C tests/driver_mcp23x17 BOARD=esp8266-esp-12x info-debug-variable-USEMODULE

auto_init auto_init_random auto_init_xtimer benchmark board boards_common_esp8266 core core_idle_thread core_init core_msg core_panic core_thread_flags cpp cpu div esp_common esp_common_periph esp_freertos esp_freertos_common esp_i2c_sw esp_idf esp_idf_esp8266 esp_idf_nvs_flash esp_idf_spi_flash esp_idf_util esp_idf_wpa_supplicant_crypto esp_sdk event event_thread event_thread_medium isrpipe log luid mcp23017 mcp23x17 mcp23x17_i2c mcp23x17_irq mtd newlib newlib_syscalls_default periph periph_common periph_cpuid periph_flash periph_gpio periph_gpio_irq periph_hwrng periph_i2c periph_i2c_sw periph_init periph_init_cpuid periph_init_flash periph_init_gpio periph_init_gpio_irq periph_init_hwrng periph_init_i2c periph_init_i2c_sw periph_init_pm periph_init_timer periph_init_uart periph_pm periph_timer periph_uart prng prng_tinymt32 ps random shell stdin stdio_uart stdio_uart_rx sys tinymt32 tsrb xtensa xtimer
```
That is, a number of specific `esp_*` modules are missing in Kconfig configuration for ESP8266 boards. The same happens for ESP32 boards.

Reply:
We still did not make it to the esp for the Kconfig modelling of dependencies, that's why it's not tested yet in the CI https://github.com/RIOT-OS/RIOT/blob/9256970517798bd844f4ee1dc4cddd9f61cef6e1/.murdock#L10-L18 (help doing so is really welcome :). It should be working fine with the platforms that already have their modules in.

Reply:
@leandrolanzieri Is there a more elegant way to map a possible choice to an enum than the following?

`driver.h`
```.c
typedef enum {
    DRIVER_MODE_1 = 0x10,
    DRIVER_MODE_2 = 0x20,
    DRIVER_MODE_3 = 0x30,
} driver_mode_t;
```
`driver_params.h`
```.c
#ifdef CONFIG_DRIVER_MODE_1
#define CONFIG_DRIVER_MODE    (DRIVER_MODE_1)
#elif CONFIG_DRIVER_MODE_2
#define CONFIG_DRIVER_MODE    (DRIVER_MODE_2)
#elif CONFIG_DRIVER_MODE_3
#define CONFIG_DRIVER_MODE    (DRIVER_MODE_3)
#endif
```

Kconfig:
```
choice
    bool "Select right mode
    default DRIVER_MODE_1

config DRIVER_MODE_1
    bool "Mode 1"

config DRIVER_MODE_2
    bool "Mode 2"

config DRIVER_MODE_3
    bool "Mode 3"

endchoice
```
When migrating default configuration parameters to `Kconfig`, a further question arose. What are the rules when a define in `driver_params.h` should be called `CONFIG_DRIVER_PARAM_<NAME>` and when a parameter should simply be called `DRIVER_PARAM_<NAME>`? I have seen a mixture which doesn't seem to be consistent.

Would be a mapping
```
#define DRIVER_PARAM_<NAME> (CONFIG_DRIVER_PARAM_<NAME>`
```
OK to keep the compatibility with existing documentation where the default parameters could be changed by overriding them by `CFLAGS` at the make command line?

Reply:
> @leandrolanzieri Is there a more elegant way to map a possible choice to an enum than the following?

What about the following:

```Kconfig
choice
    bool "Select right mode"
    default DRIVER_MODE_1

config DRIVER_MODE_1
    bool "Mode 1"

config DRIVER_MODE_2
    bool "Mode 2"

config DRIVER_MODE_3
    bool "Mode 3"

endchoice

config DRIVER_MODE
    hex
    default DRIVER_MODE_1_VAL if DRIVER_MODE_1
    default DRIVER_MODE_2_VAL if DRIVER_MODE_2
    default DRIVER_MODE_3_VAL if DRIVER_MODE_3

config DRIVER_MODE_1_VAL
    hex
    default 0x10

config DRIVER_MODE_2_VAL
    hex
    default 0x20

config DRIVER_MODE_3_VAL
    hex
    default 0x30
```
Then you can also have:
```C
typedef enum {
    DRIVER_MODE_1 = CONFIG_DRIVER_MODE_1_VAL, // not actually configurable, just not to duplicate
    DRIVER_MODE_2 = CONFIG_DRIVER_MODE_2_VAL,
    DRIVER_MODE_3 = CONFIG_DRIVER_MODE_3_VAL,
} driver_mode_t;
```
Note that this last part would only work in the case Kconfig is actually run.

---

> What are the rules when a define in `driver_params.h` should be called `CONFIG_DRIVER_PARAM_<NAME>` and when a parameter should simply be called `DRIVER_PARAM_<NAME>`? I have seen a mixture which doesn't seem to be consistent.

Personally I think many of those parameters are more a hardware configuration rather than a configuration of the driver itself, and would be better off left for other tools (e.g. the current `driver_params.h` or Device Tree). Nevertheless there are some drivers that expose these default values for their first instance as Kconfig symbols.


> Would be a mapping
> 
> ```
> #define DRIVER_PARAM_<NAME> (CONFIG_DRIVER_PARAM_<NAME>`
> ```
> 
> OK to keep the compatibility with existing documentation where the default parameters could be changed by overriding them by `CFLAGS` at the make command line?

When Kconfig is not used (it does not run by default) the `CONFIG_DRIVER_PARAM_<NAME>` value can also be injected via `CFLAGS`, so I think just a documentation update is enough, no mapping should be needed if I understood you correctly.

Reply:
> What about the following:
> ```
> ...
> config DRIVER_MODE_1_VAL
>    hex
>    default 0x10
> ...
> ```

Hm, that would require coding enumerations again, which I think is error prone. In my approach I can use the named constants as they are already defined by the driver.

> Personally I think many of those parameters are more a hardware configuration rather than a configuration of the driver itself, and would be better off left for other tools (e.g. the current `driver_params.h` or Device Tree).

I'm only talking about selections the user has to make. A typical example would be a sensor for which the desired output data rate has to be defined. The output data rate is configured using a register and all possible values are defined as an enumeration. For efficiency, the enumeration directly encodes the register values.